### PR TITLE
Add type helpers for `OnQueryStarted` callbacks

### LINKED ISSUE #4711: RTK Query 2.0 Migration: Typing `onQueryStarted` without `LifecycleApi` types
I'm working on migrating a codebase to RTK 2.0.

I frequently have a pattern where I manually update the cache. I have been doing this via the lifecycle of `onQueryStarted` as recommended in #2864. My slices looks something like this, greatly simplified:

```ts
export const usersApi = rtkQueryApiSlice.injectEndpoints({
  endpoints: (builder) => ({
    getUser: builder.query<User, void>({
      query: () => ({ name: "get-user" })
    }),
    setFoo: builder.mutation<User, string>({
     query: (payload)  => ({ name: "set-user-foo", payload }),
     onQueryStarted: updateUserOnComplete
    }),
    performBar: builder.mutation<User, string>({
     query: (payload)  => ({ name: "perform-bar", payload }),
     onQueryStarted: updateUserOnComplete
    }),
})
```

I've had those re-used `onQueryStarted` functions defined like this:

```ts
import {
  MutationLifecycleApi,
  QueryLifecycleApi,
} from "@reduxjs/toolkit/dist/query/endpointDefinitions";

const updateUserOnComplete = async (
  arg: unknown,
  {
    dispatch,
    queryFulfilled,
  }:
    | MutationLifecycleApi<unknown, BaseQueryFn, User>
    | QueryLifecycleApi<unknown, BaseQueryFn, User>
) => {
  const res = await queryFulfilled;
  dispatch(
    usersApi.util.updateQueryData("get-user", undefined, (draft) => {
      if (draft?.id === res.data.id) {
        return res.data;
      }
    })
  );
};
```

As discussed in #4554 , this type-hack isn't available any longer. What's the proper way to type this re-used `onQueryStarted` function? I've tried drilling into how the type is made from the Lifecycle:

```ts
 {
    dispatch: ThunkDispatch<unknown, unknown, UnknownAction>;
    queryFulfilled: Promise<{
      data: User;
    }>;
  }
```

But this has a type error when attempting to dispatch the action. I'm confused about that, since it seems like that's the type that `dispatch` has when declaring `onQueryStarted` inline within the query definition. 

Or maybe there's just a better way to be creating a reusable pattern here? 

Any help would be appreciated.

### PR DESCRIPTION
**Pull request by aryaemami59**
## **This PR**:

- [X] Adds the `Typed[Query|Mutation]OnQueryStarted` type helpers.
- [X] Resolves #4711.

### CODE REVIEW

File: packages/toolkit/src/query/tests/queryLifecycle.test.tsx [hunk 1]
```diff
@@ -360,7 +360,7 @@ test('query: updateCachedData', async () => {
           },
         ) {
           // calling `updateCachedData` when there is no data yet should not do anything
-          // but if there is a cache value it will be updated & overwritten by the next succesful result
+          // but if there is a cache value it will be updated & overwritten by the next successful result
```
Comment (aryaemami59) [inline packages/toolkit/src/query/tests/queryLifecycle.test.tsx:363]: Fixed this typo, hope you don't mind.

File: packages/toolkit/src/query/react/buildHooks.ts [hunk 2]
```diff
@@ -604,7 +604,7 @@ export type UseQueryStateOptions<
  * @template BaseQuery - The type of the base query function being used.
  * @template SelectedResult - The type of the selected result returned by the __`selectFromResult`__ function.
  *
- * @since 2.7.8
+ * @since 2.2.8
```
Comment (aryaemami59) [inline packages/toolkit/src/query/react/buildHooks.ts:607]: This was a typo from one of my previous PRs.

File: packages/toolkit/src/query/react/buildHooks.ts [hunk 3]
```diff
@@ -421,7 +421,7 @@ export type QueryStateSelector<
  * @template BaseQueryFunctionType - The type of the base query function being used.
  * @template SelectedResultType - The type of the selected result returned by the __`selectFromResult`__ function.
  *
- * @since 2.7.9
+ * @since 2.3.0
```
Comment (aryaemami59) [inline packages/toolkit/src/query/react/buildHooks.ts:424]: This was a typo from one of my previous PRs.

**Comment by EskiMojo14** [comment]
code wise this looks good - i'm still iffy on the name so may take a few to mull it over

**Comment by markerikson** [comment]
Per chat, shortening these to `Typed[Query|Mutation]OnQueryStarted`

**Pull request merged**